x/sophon/types/tx.go

```diff
@@ -151,7 +151,53 @@ func (m *MsgCancelOwnershipTransfer) ValidateBasic() error {
 }
 
 func (m *MsgAddUser) ValidateBasic() error {
-	return fmt.Errorf("not implemented")
+	if _, err := sdk.AccAddressFromBech32(m.AdminAddress); err != nil {
+		return sdkerrors.ErrInvalidAddress.Wrapf("invalid admin address: %s", m.AdminAddress)
+	}
+
+	if len(m.SophonPublicKey) == 0 {
+		return sdkerrors.ErrInvalidRequest.Wrap("empty public key")
+	}
+
+	_, err := hex.DecodeString(m.SophonPublicKey)
+	if err != nil {
+		return sdkerrors.ErrInvalidRequest.Wrapf("invalid hex in pubkey: %s", m.SophonPublicKey)
+	}
+
+	if len(m.UserId) == 0 {
```

Comment:
Shouldn't we add some additional constraints on the user ID?

Reply:
Good one. I went with 66 as most of our public key identifiers are hex strings of that length.